File: PolishNotation.java
/**
 *
 * @author SWAPNIL
 * 06-Oct-2016
 *
 *@additionalAuthor InetZen
 *16-May-2018
 */

import java.util.*;
import java.io.*;

public class PolishNotation {

    /**
     * @param args the command line arguments
     */
    
    final ArrayList operators = new ArrayList(Arrays.asList('+','-','*','/','^'));
    final Map <String,Integer> OperatorPriority = new HashMap <String,Integer> () {{
        put("+",0);
        put("-",0);
        put("*",1);
        put("/",1);
        put("^",2);
    }};
    

    public static void main(String[] args) {
        PolishNotation polish = new PolishNotation();
        polish.calculator();
    } 
    
    public void calculator () {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("Escape Character is ']'. Enter your input at the prompt");
        String input;
        try {
            while(!(input = in.readLine()).equals("]")) {
              LinkedList <String> Infix = toInfix(input);
              LinkedList <String> PostfixExpr = toPostfix(Infix);
              System.out.println(evaluatePostfix(PostfixExpr));
            }
        } catch(IOException ex) {ex.printStackTrace();}
    }
    
    private LinkedList toInfix (String expr) {
        LinkedList <String> infixExpression = new LinkedList<>();
        for(int i =0; i < expr.length(); i++) {
            char current = expr.charAt(i);
            if(current == ' ');
            else if(isNum(current)) {
                int numLen = getNumber(expr.substring(i));
                infixExpression.add(expr.substring(i, i+numLen));
                i += numLen - 1;
            }
            else if(isOperator(current)) {
                infixExpression.add(String.valueOf(current));
            }
            else if(current == '(' || current == ')') {
                infixExpression.add(String.valueOf(current));
            }
            else {
                System.err.println("Non Integer value Entered. Fatal Error Program Terminated!");
                System.exit(0);
            }
        }
        infixExpression.add(")");
        return infixExpression;
    }
    
    private LinkedList <String> toPostfix(LinkedList <String> infixExpr) {
        Stack <String> infixOperatorStack = new Stack <> ();
        LinkedList <String> PostfixExpr = new LinkedList <> ();
        
        infixOperatorStack.push("(");  //1. Push '(' to the Stack
        while(!infixExpr.isEmpty()) {
            if(isNum(infixExpr.peek()))
                PostfixExpr.add(infixExpr.remove());
            else if(infixExpr.peek().equals("("))
                infixOperatorStack.push(infixExpr.remove());
            else if(infixExpr.peek().length() == 1 && isOperator(infixExpr.peek().charAt(0))) {
                if(infixOperatorStack.peek().equals("(")) {
                    infixOperatorStack.push(infixExpr.remove());
                }
                else {
                    while(OperatorPriority.get(infixExpr.peek()) <= OperatorPriority.get(infixOperatorStack.peek())) {
                        PostfixExpr.add(infixOperatorStack.pop());
                        if(infixOperatorStack.peek().equals("(")) break;
                    }
                    infixOperatorStack.push(infixExpr.remove());
                }
            }
            else if(infixExpr.peek().equals(")")) {
                while(!infixOperatorStack.peek().equals("(")) {
                    PostfixExpr.add(infixOperatorStack.pop());
                }
                infixOperatorStack.pop();
                infixExpr.remove();
            }            
        }
        return PostfixExpr;
    }
    
    private int evaluatePostfix(LinkedList <String> postfixExpr) {
        
        Stack <String> postfixOperandStack = new Stack <> ();
        
        postfixExpr.add(")");
        while(!postfixExpr.peek().equals(")")) {
            if(isNum(postfixExpr.peek())) {
                postfixOperandStack.push(postfixExpr.remove());
            }
            else if (postfixExpr.peek().length() == 1 && isOperator(postfixExpr.peek().charAt(0))) {
                int result = Evaluate(postfixOperandStack.pop(),postfixOperandStack.pop(),postfixExpr.remove().charAt(0));
                postfixOperandStack.push(String.valueOf(result));
            }
        } 
        return Integer.parseInt(postfixOperandStack.pop());
    }
    
    private boolean isNum (char ch) {
        return ch >= '0' && ch <= '9';
    }
    
    private boolean isNum (String num) {
        for(int i = 0; i < num.length(); i++) {
            if(!isNum(num.charAt(i)))
                return false;
        }
        return true;
    }
    
    private boolean isOperator(char ch) {
        return operators.contains(ch);
    }
    
    private int getNumber(String sliceExpr) {
        int i = 0;
        while(i < sliceExpr.length() && isNum(sliceExpr.charAt(i))) 
            i++;
        return i;
    }
    
    private int Evaluate (String first, String second, char operator) {
        int result = 0;
        int A = Integer.parseInt(first);
        int B = Integer.parseInt(second);
        switch(operator) {
            case '+' :
                result = B + A;
                break;
            case '-':
                result = B - A;
                break;
            case '*':
                result = B * A;
                break;
            case '/':
                result = B / A;
                break;
            case '^':
                double resultPow = Math.pow(B,A);
                String imd = String.valueOf(resultPow);
                String split[] = imd.split("\\.");
                result = Integer.parseInt(split[0]);
                break;
        }
        return result;
    }
}

File: PolishNotation.js
/*
* 
* @author github.com/SwapnilB31
* 06-Oct-2016
*
* Change on 01-Feb-2021
*
* This program is written within the specifications of Nashorn Javascript. Nashorn is a Javascript Engine written in Java
* that ships with the JDK since Java 8.
*
* To run this program, invoke it with the jjs application that you'll find in your jdk/bin
*/

var operators = ["+","-","*","/","^"];

var operatorPrecedence = {
  "+" : 0,
  "-" : 0,
  "*" : 1,
  "/" : 1,
  "^" : 2
};


function isNum(chunk) {
    for(var i in chunk)
      if(!(isNumeral(chunk[i]) || chunk[i] === '.'))
        return false;
      return true;
}

function isNumeral(ele) {
  return ele >= '0' && ele <= '9';
}

function isOperator(ele) {
  for(var i in operators)
    if(ele === operators[i])
      return true;
    return false;
}

function getNumber(sliceExpr) {
  var i = 0;
  while( i < sliceExpr.length && (isNumeral(sliceExpr[i]) || sliceExpr[i] === "."))
    i++;
  return i;
}

function Evaluate (A,B,op) {
  var result = 0;
  switch (op) {
    case "+":
      result = B + A;
      break;
    case "-":
        result = B - A;
        break;
    case "*":
      result = B * A;
      break;
    case "/":
      result = B / A;
      break;
    case "^":
      result = Math.pow(B,A);
  }
  return result;
}

function toInfix (expr) {
  var infixExpression = [];
  for(var i = 0; i < expr.length; i++) {
    if (expr[i] === " ");
    else if (isNumeral(expr[i])) {
      var numLen = getNumber(expr.slice(i));
      infixExpression.push(expr.slice(i,i+numLen));
      i += numLen - 1;
    }
    else if (isOperator(expr[i])) {
      infixExpression.push(expr[i]);
    }
    else if (expr[i] === "(" || expr[i] === ")") {
      infixExpression.push(expr[i]);
    }
    else {
      return false;
    }
  }
  infixExpression.push(")");
  return infixExpression;
}

function toPostfix(infixExpr) {
  var infixOperatorStack = ["("];
  var postfixExpr = [];
  while(infixOperatorStack.length > 0) {
    if (isNum(infixExpr[0])) {
      postfixExpr.push(infixExpr.shift());
    }
    else if (infixExpr[0] === "(") {
      infixOperatorStack.push(infixExpr.shift());
    }
    else if (isOperator(infixExpr[0])) {
      if(infixOperatorStack[infixOperatorStack.length - 1] === "(")
        infixOperatorStack.push(infixExpr.shift());
      else {
        while (operatorPrecedence[infixOperatorStack[infixOperatorStack.length - 1]] >= operatorPrecedence[infixExpr[0]]) {
          postfixExpr.push(infixOperatorStack.pop());
          if(infixOperatorStack[infixOperatorStack.length - 1] === "(") break;
        }
        infixOperatorStack.push(infixExpr.shift());
      }
    }
    else if (infixExpr[0] === ")") {
      while(infixOperatorStack[infixOperatorStack.length - 1] !== "(") {
        postfixExpr.push(infixOperatorStack.pop());
      }
      infixOperatorStack.pop();
      infixExpr.shift();
    }
  }
  return postfixExpr;
}

function evaluatePostfix(postfixExpr) {
  var postfixOperandStack = [];
  postfixExpr.push(")");
  while(postfixExpr[0] !== ")") {
    if(isNum(postfixExpr[0])) {
      postfixOperandStack.push(postfixExpr.shift());
    }
    else if (isOperator(postfixExpr[0])) {
      var result = Evaluate(parseFloat(postfixOperandStack.pop()),parseFloat(postfixOperandStack.pop()),postfixExpr.shift());
      postfixOperandStack.push(result.toString());
    }
  }
  return postfixOperandStack.pop();
}

function calculator() {
  var BufferedReader = Java.type("java.io.BufferedReader");
  var InputStreamReader = Java.type("java.io.InputStreamReader");
  var System = Java.type("java.lang.System");
  var inp = new BufferedReader(new InputStreamReader(System.in));
  try {
    print("The Escape character is ']'. Enter you expression at the prompt");
    var input = inp.readLine();
    while(input != "]") {
      var infix = toInfix(input);
      if(infix === false) {
        print("Non-numerical value entered. Fatal Error Program Terminated!");
        return;
      }
      var postfix = toPostfix(infix);
      print(evaluatePostfix(postfix));
      input = inp.readLine();
    }
  } catch (e) {
    e.printStackTrace();
  }
}

calculator();
